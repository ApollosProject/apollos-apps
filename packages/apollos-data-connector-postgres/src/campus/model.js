import { DataTypes } from 'sequelize';
import { defineModel, configureModel } from '../postgres';

const createModel = defineModel({
  modelName: 'campus',
  resolveType: () => 'Campus',
  external: true,
  attributes: {
    name: DataTypes.TEXT,
    street1: DataTypes.TEXT,
    street2: DataTypes.TEXT,
    city: DataTypes.TEXT,
    state: DataTypes.TEXT,
    postalCode: DataTypes.TEXT,
    latitude: DataTypes.FLOAT,
    longitude: DataTypes.FLOAT,
    imageUrl: DataTypes.TEXT,
    digital: DataTypes.BOOLEAN,
  },
});

const setupModel = configureModel(({ sequelize }) => {
  sequelize.models.campus.hasMany(sequelize.models.people);
  sequelize.models.people.belongsTo(sequelize.models.campus);
});

export { createModel, setupModel };
