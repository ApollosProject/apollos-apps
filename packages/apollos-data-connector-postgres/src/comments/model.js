import { DataTypes } from 'sequelize';
import { defineModel, configureModel } from '../postgres';

const Visibility = {
  PUBLIC: 'PUBLIC',
  PRIVATE: 'PRIVATE',
  FOLLOWERS: 'FOLLOWERS',
};

const createModel = defineModel({
  modelName: 'comments',
  resolveType: () => 'Comment',
  attributes: {
    text: DataTypes.TEXT,
    visibility: DataTypes.ENUM(Object.values(Visibility)),
    externalParentId: DataTypes.TEXT,
    externalParentType: DataTypes.TEXT,
    externalParentSource: DataTypes.TEXT,
    flagCount: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
    },
  },
});

const setupModel = configureModel(({ sequelize }) => {
  sequelize.models.comments.belongsTo(sequelize.models.people);
  sequelize.models.people.hasMany(sequelize.models.comments);

  // Used to identify user relationships when querying for comments
  // Specifically, to sort the comments of those you follow to the top
  sequelize.models.comments.hasMany(sequelize.models.follows, {
    foreignKey: 'followedPersonId',
    sourceKey: 'personId',
    constraints: false,
  });
});

export { createModel, Visibility, setupModel };
