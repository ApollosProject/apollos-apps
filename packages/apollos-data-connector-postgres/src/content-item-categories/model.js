import { DataTypes } from 'sequelize';
import { defineModel, configureModel } from '../postgres';

const createModel = defineModel({
  modelName: 'contentItemCategory',
  resolveType: () => 'ContentItemCategory',
  external: true,
  attributes: {
    title: DataTypes.TEXT,
  },
});

const setupModel = configureModel(({ sequelize }) => {
  sequelize.models.contentItem.belongsTo(sequelize.models.contentItemCategory);
  sequelize.models.contentItemCategory.hasMany(sequelize.models.contentItem);
});

export { createModel, setupModel };
