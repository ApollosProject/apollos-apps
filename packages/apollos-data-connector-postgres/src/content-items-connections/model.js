import { DataTypes } from 'sequelize';
import { defineModel, configureModel } from '../postgres';

const createModel = defineModel({
  modelName: 'contentItemsConnection',
  resolveType: () => 'ContentItemConnection', // shouldn't get called, ContentItem will be set on the shovel.
  external: true,
  attributes: {
    order: DataTypes.INTEGER,
  },
});

const setupModel = configureModel(({ sequelize }) => {
  sequelize.models.contentItem.belongsToMany(sequelize.models.contentItem, {
    through: sequelize.models.contentItemsConnection,
    as: 'children',
    foreignKey: 'parentId',
    otherKey: 'childId',
  });

  sequelize.models.contentItem.belongsToMany(sequelize.models.contentItem, {
    through: sequelize.models.contentItemsConnection,
    as: 'parents',
    otherKey: 'parentId',
    foreignKey: 'childId',
  });
});

export { createModel, setupModel };
