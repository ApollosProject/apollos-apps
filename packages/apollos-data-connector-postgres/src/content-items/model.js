import { DataTypes } from 'sequelize';
import { defineModel, configureModel } from '../postgres';

const createModel = defineModel({
  modelName: 'contentItem',
  resolveType: () => 'UniversalContentItem', // shouldn't get called, ContentItem will be set on the shovel.
  external: true,
  attributes: {
    title: DataTypes.TEXT,
    summary: DataTypes.TEXT,
    htmlContent: DataTypes.TEXT,
    publishAt: DataTypes.DATE,
    active: DataTypes.BOOLEAN,
  },
});

const setupModel = configureModel(({ sequelize }) => {
  sequelize.models.contentItem.belongsTo(sequelize.models.media, {
    as: 'coverImage',
    foreignKey: 'coverImageId',
  });

  sequelize.models.contentItem.belongsTo(sequelize.models.contentItem, {
    as: 'parent',
    foreignKey: 'parentId',
  });
});

export { createModel, setupModel };
