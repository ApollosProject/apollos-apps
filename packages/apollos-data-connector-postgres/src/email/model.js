const createModel = () => {};

const setupModel = () => {};

export { createModel, setupModel };
