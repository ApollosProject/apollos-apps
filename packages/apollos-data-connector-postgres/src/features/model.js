import { DataTypes } from 'sequelize';
import { defineModel, configureModel } from '../postgres';

const createModel = defineModel({
  modelName: 'feature',
  resolveType: () => 'Feature',
  attributes: {
    parentId: DataTypes.UUID,
    parentType: DataTypes.TEXT,
    data: DataTypes.JSONB,
    type: DataTypes.TEXT,
  },
  sequelizeOptions: {
    tableName: 'feature',
    underscored: true,
  },
});

const setupModel = configureModel(({ sequelize }) => {
  sequelize.models.feature.belongsTo(sequelize.models.contentItem, {
    foreignKey: 'parentId',
  });
  sequelize.models.contentItem.hasMany(sequelize.models.feature, {
    foreignKey: 'parentId',
    scope: { parentType: 'ContentItem' },
  });
});

export { createModel, setupModel };
