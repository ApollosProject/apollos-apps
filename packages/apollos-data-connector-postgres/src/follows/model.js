import { DataTypes } from 'sequelize';
import { configureModel, defineModel } from '../postgres';

const FollowState = {
  REQUESTED: 'REQUESTED',
  DECLINED: 'DECLINED',
  ACCEPTED: 'ACCEPTED',
};

const createModel = defineModel({
  modelName: 'follows',
  resolveType: () => 'Follow',
  attributes: {
    requestPersonId: DataTypes.UUID,
    followedPersonId: DataTypes.UUID,
    state: DataTypes.ENUM(Object.values(FollowState)),
  },
});

const setupModel = configureModel(({ sequelize }) => {
  sequelize.models.follows.belongsTo(sequelize.models.people, {
    foreignKey: 'requestPersonId',
    as: 'follower',
  });
  sequelize.models.follows.belongsTo(sequelize.models.people, {
    foreignKey: 'followedPersonId',
    as: 'following',
  });
  sequelize.models.people.hasMany(sequelize.models.follows, {
    foreignKey: 'requestPersonId',
    as: 'requestedFollows',
  });
  sequelize.models.people.hasMany(sequelize.models.follows, {
    foreignKey: 'followedPersonId',
    as: 'followingRequests',
  });

  sequelize.models.people.belongsToMany(sequelize.models.people, {
    through: sequelize.models.follows,
    foreignKey: 'followedPersonId',
    otherKey: 'requestPersonId',
    as: 'followers',
  });
});

export { createModel, setupModel, FollowState };
