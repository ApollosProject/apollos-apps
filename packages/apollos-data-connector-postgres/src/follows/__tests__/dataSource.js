import ApollosConfig from '@apollosproject/config';
import { sequelize, sync } from '../../postgres/index';
import { createModel, FollowState, setupModel } from '../model';
import { createModel as createPeopleModel } from '../../people/model';
import {
  createModel as createCampusModel,
  setupModel as setupCampusModel,
} from '../../campus/model';
import FollowDataSource from '../dataSource';
import PeopleDataSource from '../../people/dataSource';

let currentPersonId = 1;

const context = {
  dataSources: {
    Auth: {
      getCurrentPerson: () => ({
        id: currentPersonId,
      }),
    },
    Person: {
      resolveId: (id) => id,
      whereCurrentPerson: ({ id }) => ({ id }),
    },
  },
};

let person1;
let person2;
let person3;
let person4;

describe('Apollos Postgres FollowRequest DataSource', () => {
  // let followDataSource;
  beforeEach(async () => {
    try {
      await createPeopleModel();
      await createModel();
      await setupModel();
      await createCampusModel();
      await setupCampusModel();
      await sync({ force: true });

      // Make sure people exist for all of our test ids
      const peopleDataSource = new PeopleDataSource();
      peopleDataSource.initialize({ context });
    } catch (e) {
      console.error(e);
    }
    person1 = await sequelize.models.people.create({
      originId: '11',
      originType: 'rock',
    });
    person2 = await sequelize.models.people.create({
      originId: '22',
      originType: 'rock',
    });
    person3 = await sequelize.models.people.create({
      originId: '33',
      originType: 'rock',
    });
    person4 = await sequelize.models.people.create({
      originId: '44',
      originType: 'rock',
    });
    currentPersonId = person1.id;
  });
  afterEach(async () => {
    await sequelize.drop({});
    currentPersonId = 1;
  });

  it('should create new follow request', async () => {
    const followDataSource = new FollowDataSource();

    followDataSource.initialize({ context });

    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    const follows = await followDataSource.model.findAll({
      where: {
        followedPersonId: person2.id,
      },
    });

    expect(follows.length).toBe(1);
    expect(follows[0].state).toBe(FollowState.REQUESTED);
  });

  it('should ignore existing unaccepted request', async () => {
    const followDataSource = new FollowDataSource();

    followDataSource.initialize({ context });

    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    // Send another request
    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    const follows = await followDataSource.model.findAll({
      where: {
        followedPersonId: person2.id,
      },
    });

    expect(follows.length).toBe(1);
    expect(follows[0].state).toBe(FollowState.REQUESTED);
  });

  it('should ignore request', async () => {
    const followDataSource = new FollowDataSource();

    followDataSource.initialize({ context });

    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    // Find and deny the request
    let existingRequest = await followDataSource.model.findOne({
      where: {
        followedPersonId: person2.id,
      },
    });

    currentPersonId = person2.id;

    const ignoreResult = await followDataSource.ignoreFollowRequest({
      requestPersonId: `Person:${person1.id}`,
    });

    existingRequest = await followDataSource.model.findOne({
      where: {
        followedPersonId: person2.id,
      },
    });

    expect(existingRequest.state).toBe(FollowState.DECLINED);
    expect(existingRequest.id).toBe(ignoreResult.followId);
  });

  it('should accept request', async () => {
    const followDataSource = new FollowDataSource();

    followDataSource.initialize({ context });

    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    // Find and accept the request
    let existingRequest = await followDataSource.model.findOne({
      where: {
        followedPersonId: person2.id,
      },
    });

    currentPersonId = person2.id;

    const acceptResult = await followDataSource.acceptFollowRequest({
      requestPersonId: `Person:${person1.id}`,
    });

    existingRequest = await followDataSource.model.findOne({
      where: {
        followedPersonId: person2.id,
      },
    });

    expect(existingRequest.state).toBe(FollowState.ACCEPTED);
    expect(existingRequest.id).toBe(acceptResult.followId);
  });

  it('should ignore existing accepted request', async () => {
    const followDataSource = new FollowDataSource();

    followDataSource.initialize({ context });

    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    // Find and accept the request
    await followDataSource.model.findOne({
      where: {
        followedPersonId: person2.id,
      },
    });

    currentPersonId = person2.id;

    await followDataSource.acceptFollowRequest({
      requestPersonId: `Person:${person1.id}`,
    });

    currentPersonId = person1.id;

    // Send another request
    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    const follows = await followDataSource.model.findAll({
      where: {
        followedPersonId: person2.id,
      },
    });

    expect(follows.length).toBe(1);
    expect(follows[0].state).toBe(FollowState.ACCEPTED);
  });

  it('should reset existing denied request', async () => {
    const followDataSource = new FollowDataSource();

    followDataSource.initialize({ context });

    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    // Find and deny the request
    await followDataSource.model.findOne({
      where: {
        followedPersonId: person2.id,
      },
    });

    currentPersonId = person2.id;

    await followDataSource.ignoreFollowRequest({
      requestPersonId: `Person:${person1.id}`,
    });

    currentPersonId = person1.id;

    // Send another request
    await followDataSource.requestFollow({
      followedPersonId: `Person:${person2.id}`,
    });

    const follows = await followDataSource.model.findAll({
      where: {
        followedPersonId: person2.id,
      },
    });

    expect(follows.length).toBe(1);
    expect(follows[0].state).toBe(FollowState.REQUESTED);
  });

  it("should get a user's list of sugggested people to follow", async () => {
    const followDataSource = new FollowDataSource();
    followDataSource.initialize({ context });
    // Lengthy setup :g
    const mainCampus = await sequelize.models.campus.create({
      name: 'Main Campus',
      originId: '1',
      originType: 'rock',
    });
    const satCampus = await sequelize.models.campus.create({
      name: 'Satilite Campus',
      originId: '2',
      originType: 'rock',
    });
    await sequelize.models.people.create({
      originId: '1',
      originType: 'rock',
      firstName: 'Jim',
      lastName: 'Bob',
      email: 'jim@bob.com',
      campusId: satCampus.id,
    });
    await sequelize.models.people.create({
      originId: '2',
      originType: 'rock',
      firstName: 'Vincent',
      lastName: 'Wilson',
      email: 'vin@wil.com',
      campusId: mainCampus.id,
    });
    await sequelize.models.people.create({
      originId: '3',
      originType: 'rock',
      firstName: 'Nick',
      lastName: 'Offerman',
      email: 'nick@offer.man',
    });
    ApollosConfig.loadJs({
      SUGGESTED_FOLLOWS: [
        'nick@offer.man',
        {
          email: 'vin@wil.com',
          campusId: mainCampus.id,
        },
        {
          email: 'jim@bob.com',
          campusId: satCampus.id,
        },
      ],
    });

    const me = await sequelize.models.people.create({
      originId: '4',
      originType: 'rock',
      firstName: 'Me',
      lastName: 'Myself',
      campusId: satCampus.id,
    });

    const suggestedFollowers = await followDataSource.getStaticSuggestedFollowsFor(
      me
    );

    expect(suggestedFollowers.map(({ email }) => email)).toEqual([
      'jim@bob.com',
      'nick@offer.man',
    ]);
  });

  it("should get a user's list of sugggested people to follow if they have no campus", async () => {
    const followDataSource = new FollowDataSource();
    followDataSource.initialize({ context });
    // Lengthy setup :g
    const mainCampus = await sequelize.models.campus.create({
      name: 'Main Campus',
      originId: '1',
      originType: 'rock',
    });
    const satCampus = await sequelize.models.campus.create({
      name: 'Satilite Campus',
      originId: '2',
      originType: 'rock',
    });
    await sequelize.models.people.create({
      originId: '1',
      originType: 'rock',
      firstName: 'Jim',
      lastName: 'Bob',
      email: 'jim@bob.com',
      campusId: satCampus.id,
    });
    await sequelize.models.people.create({
      originId: '2',
      originType: 'rock',
      firstName: 'Vincent',
      lastName: 'Wilson',
      email: 'vin@wil.com',
      campusId: mainCampus.id,
    });
    await sequelize.models.people.create({
      originId: '3',
      originType: 'rock',
      firstName: 'Nick',
      lastName: 'Offerman',
      email: 'nick@offer.man',
    });
    ApollosConfig.loadJs({
      SUGGESTED_FOLLOWS: [
        'nick@offer.man',
        {
          email: 'vin@wil.com',
          campusId: mainCampus.id,
        },
        {
          email: 'jim@bob.com',
          campusId: satCampus.id,
        },
      ],
    });

    const me = await sequelize.models.people.create({
      originId: '4',
      originType: 'rock',
      firstName: 'Me',
      lastName: 'Myself',
      campusId: null,
    });

    const suggestedFollowers = await followDataSource.getStaticSuggestedFollowsFor(
      me
    );

    expect(suggestedFollowers.map(({ email }) => email)).toEqual([
      'nick@offer.man',
    ]);
  });

  it('should not suggest a user follows themselves', async () => {
    const followDataSource = new FollowDataSource();
    followDataSource.initialize({ context });
    const me = await sequelize.models.people.create({
      originId: '4',
      originType: 'rock',
      firstName: 'Me',
      lastName: 'Myself',
      email: 'me@me.com',
    });
    ApollosConfig.loadJs({
      SUGGESTED_FOLLOWS: ['me@me.com'],
    });

    currentPersonId = me.id;

    const suggestedFollowers = await followDataSource.getStaticSuggestedFollowsFor(
      me
    );

    expect(suggestedFollowers.map(({ email }) => email)).toEqual([]);
  });
  it('should throw an error when passing a non-uuid to getStaticSuggestedFollowsFor', async () => {
    const followDataSource = new FollowDataSource();
    followDataSource.initialize({ context });
    const invalidCampus = followDataSource.getStaticSuggestedFollowsFor({
      campusId: 1,
    });
    await expect(invalidCampus).rejects.toMatchSnapshot();

    const invalidId = followDataSource.getStaticSuggestedFollowsFor({ id: 1 });
    await expect(invalidId).rejects.toMatchSnapshot();
  });

  it('should not suggest a user follows someone they already follow', async () => {
    const followDataSource = new FollowDataSource();
    followDataSource.initialize({ context });
    const me = await sequelize.models.people.create({
      originId: '4',
      originType: 'rock',
      firstName: 'Me',
      lastName: 'Myself',
      email: 'me@me.com',
    });
    const followedAlready = await sequelize.models.people.create({
      originId: '5',
      originType: 'rock',
      firstName: 'Me',
      lastName: 'Myself',
      email: 'followed@already.com',
    });

    await sequelize.models.follows.create({
      requestPersonId: me.id,
      followedPersonId: followedAlready.id,
      state: 'ACCEPTED',
    });

    ApollosConfig.loadJs({
      SUGGESTED_FOLLOWS: ['followed@already.com'],
    });

    currentPersonId = me.id;

    const suggestedFollowers = await followDataSource.getStaticSuggestedFollowsFor(
      me
    );

    expect(suggestedFollowers).toEqual([]);
  });
  it('should return list of users requesting to follow the current user', async () => {
    const followDataSource = new FollowDataSource();

    followDataSource.initialize({ context });

    // Have several people follow user 1
    await Promise.all(
      [person2, person3, person4].map(({ id }) => {
        currentPersonId = id;
        return followDataSource.requestFollow({
          followedPersonId: `Person:${person1.id}`,
        });
      })
    );

    currentPersonId = person1.id;

    const follows = await followDataSource.followRequests();

    expect(follows.length).toBe(3);
    expect(follows.map((f) => f.id).includes(person1.id)).toBeFalsy();
  });

  it('should only return unanswered requests', async () => {
    const followDataSource = new FollowDataSource();

    followDataSource.initialize({ context });

    // Have several people follow user 1
    const followRange = [person2, person3, person4];
    await Promise.all(
      followRange.map(({ id }) => {
        currentPersonId = id;
        return followDataSource.requestFollow({
          followedPersonId: `Person:${person1.id}`,
        });
      })
    );

    currentPersonId = person1.id;

    // accept one of the requests
    await followDataSource.acceptFollowRequest({
      requestPersonId: `Person:${person2.id}`,
    });

    const follows = await followDataSource.followRequests();

    // Should be one fewer request
    expect(follows.length).toBe(followRange.length - 1);
  });
});
