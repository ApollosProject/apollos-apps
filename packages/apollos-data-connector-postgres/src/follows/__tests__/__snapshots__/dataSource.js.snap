// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apollos Postgres FollowRequest DataSource should throw an error when passing a non-uuid to getStaticSuggestedFollowsFor 1`] = `[Error: ID 1 is not a valid UUID. You are probably passing a Rock (or other) id to getStaticSuggestedFollowsFor when it expects a Postgres UUID.]`;

exports[`Apollos Postgres FollowRequest DataSource should throw an error when passing a non-uuid to getStaticSuggestedFollowsFor 2`] = `[Error: ID 1 is not a valid UUID. You are probably passing a Rock (or other) id to getStaticSuggestedFollowsFor when it expects a Postgres UUID.]`;
